### PR TITLE
Add documentation for jax.jvp and jax.vjp.

### PR DESCRIPTION
**Pull request by hawkinsp**
Please do actually read this one, I'm not up with my autodiff terminology/math.
(#61 )

### CODE REVIEW

File: jax/api.py [hunk 1]
```diff
@@ -410,6 +410,29 @@ def papply_fun(*args, **kwargs):
...
+
+  Args:
+    fun: Function to be differentiated. Its arguments should be arrays, scalars,
+      or standard Python containers of arrays or scalars. It should return an
+      array, scalar, or standard Python container of arrays or scalars.
+    primals: The primal values at which the Jacobian of `fun` should be
+      evaluated. Should be a tuple of arrays, scalar, or standard Python
+      container thereof.
+    tangents: The tangent vector for which the Jacobian-vector product should be
+      evaluated. Should be a tuple of arrays, scalar, or standard Python
```
Comment (mattjj) [inline jax/api.py]: Could add that tangents should have the same shape as primals, and that the output is a tangent vector with the same shape as primals_out. 

Could also add that the length of primals and tangents is equal to the number of formal parameters to fun.

File: jax/api.py [hunk 2]
```diff
@@ -443,6 +466,31 @@ def fun(*args):
...
+      or standard Python containers of arrays or scalars. It should return an
+      array, scalar, or standard Python container of arrays or scalars.
+    primals: A sequence of of primal values at which the Jacobian of `fun`
+      should be evaluated. Should be a tuple of arrays, scalar, or standard
+      Python containers thereof.
+
+  Returns:
+    A `(primals_out, gradient)` pair, where `primals_out` is `fun(*primals)`.
+    `gradient` is a function from tangent values that computes vector-Jacobian
+    product of `fun` in an epsilon ball around `primals`.
```
Comment (mattjj) [inline jax/api.py]: Could add that the gradient fun (maybe call it vjp_at_primals?) accepts an argument cotangent vector with the same shape as primals_out. 

Instead if “in an epsilon ball around” I think we should just write “at”. Could also say it returns a cotangent vector with the same shape as primals.

**Review by mattjj — APPROVED** [review]
LGTM! Some tweaks suggested.

File: jax/api.py [hunk 3]
```diff
@@ -410,6 +410,29 @@ def papply_fun(*args, **kwargs):
...
+      array, scalar, or standard Python container of arrays or scalars.
+    primals: The primal values at which the Jacobian of `fun` should be
+      evaluated. Should be a tuple of arrays, scalar, or standard Python
+      container thereof.
+    tangents: The tangent vector for which the Jacobian-vector product should be
+      evaluated. Should be a tuple of arrays, scalar, or standard Python
+      container thereof.
+
+  Returns:
+    A `(primals_out, tangents_out)` pair, where `primals_out` is `fun(primals)`,
```
Comment (mattjj) [inline jax/api.py]: Typo: `fun(primals)` should be `fun(*primals)`.

**Review by mattjj — APPROVED** [review]

File: jax/api.py (hunk 2 above)
Comment (mattjj) [reply jax/api.py]: Maybe instead of "tangent values" say "cotangent values". We readily identify all these things but that language can be a kind of type hint to the careful reader.

Comment (mattjj) [reply jax/api.py]: Taking those suggestions together, we might end up with:

A `(primals_out, vjpfun)` pair, where `primals_out` is `fun(*primals)`. `vjpfun` is a function from a cotangent vector with the same shape as `primals_out` to a tuple of cotangent vectors with the same shape as `primals`, representing the vector-Jacobian product of `fun` evaluated at `primals`.

File: jax/api.py [hunk 4]
```diff
@@ -443,6 +466,31 @@ def fun(*args):
   return apply_jaxtree_fun(fun, io_tree, *py_args)
 
 def vjp(fun, *primals):
+  """Compute a (reverse-mode) vector-Jacobian product of `fun`.
+
+  This is a more general form of `grad` that can be used for functions with
+  non-scalar outputs. For most common use cases, you most likely want `grad`
```
Comment (mattjj) [inline jax/api.py]: Maybe delete this paragraph. It's true that `vjp` can handle non-scalar outputs, but so can `jacobian`. And this paragraph leaves off that you might want to use this to control the contangent vector that is pulled back.

I don't think any users will mistakenly use this function over `grad`.

An alternative way to convey this information might be to write something like:

`grad` is implemented as a special case of `vjp`, rougly as:
```python
def grad(f):
  def gradfun(*args):
    _, vjpfun = vjp(f, *args)
    return vjpfun(1.)
  return gradfun
```

Though I'm not sure that's necessary either.

**Comment by hawkinsp** [comment]
All done.
